### ISSUE TITLE
[Bug?] Difference in behavior of `from_type` between Union type and Union class field

### ISSUE THREAD

**Issue by segevmalool** [HypothesisWorks/hypothesis#2946]
Here's an example of the issue:

```python
from pydantic import BaseModel
from typing import Union
from hypothesis import given
from hypothesis.strategies import from_type

class Orange(BaseModel):
    pass

class Apple(BaseModel):
    pass


FruitUnion = Union[Orange, Apple]

class SomeFruit(BaseModel):
    fruit: FruitUnion


# Balances the types of fruits
@given(from_type(FruitUnion))
def testFruitUnion(x: FruitUnion):
    print(type(x))

# Always outputs Orange
@given(from_type(SomeFruit))
def testSomeFruit(x: SomeFruit):
    print(x)

if __name__ == '__main__':
    testFruitUnion()
    testSomeFruit()
```

This causes a lack of coverage when generating test data.

**Comment by Zac-HD** [comment]
This seems to come down to a weird divergence where we infer the correct strategy for the `SomeFruit.fruit` field, but then don't generate it.  Unsure what's going on exactly, but I *think* it's something about introspecting Pydantic models.  I'm pretty sure #2940 is a distinct but related problem.

Needs: more investigation.  I'll get to it eventually but external contributions most welcome 🙂

**Comment by Zac-HD** [comment]
Closing this because it seems unlikely that anyone will fix it before the release of Pydantic 2.0, which is expected to fix this.
